_episodes/05-Containers.md

```diff
@@ -3,8 +3,462 @@ title: "Containers"
+
+There are two ways to go about this:
+1. You can copy a file from your local machine into a (running) docker container using `docker cp [OPTIONS] CONTAINER:SRC_PATH DEST_PATH`
+2. You can mount part of your local file system to a location within the container using `docker run --mount [OPTIONS] CONTAINER`
+
+The first option can also be used to copy files *from* the container to your local machine.
+
+The second option is often best when you are developing code that needs to run within the container because you can use your favorite IDE on your local machine to work on the files, and then run them within the container without having to constantly copy things back and forth.
+
+> ## Mount a local directory into your container
+>  On your machine, `touch` a new file called `do_things.py` and add the following:
+> ~~~
+> #! /usr/bin/env python
+> import socket
+> print(f"Hello from {socket.gethostname()}")
+> ~~~
+> {: .language-python}
+>
+> Now run:
+> ~~~
+> docker run -it --mount type=bind,source="$(pwd)",target=/app python:3.8.5 bash
+> ~~~
+> {: .language-bash}
+> Navigate to the `/app` directory and run `python do_stuff.py`
+{: .challenge}
+
+In the above exercise you should see that the host name is some random string of letters and numbers, which is (hopefully) different from your local machines name.
+Note that when we are working inside the container we are working as `root`, which means we have *all the privileges*.
+If you bind a local path to one inside the container, then you'll have `root` access to that path.
+So it's a good idea **not** to mount `/` inside the container!
+Being `root` user inside the container also means that any files which you create in the mounted directory will be owned by root on your local machine.
+Having root privileges within the container is a big reason why you wont see docker being provided on an HPC.
+
+
+## Building (docker) containers
+As well as using pre-made containers, you can build your own.
+Whilst it's possible to build a container from scratch, it's recommended that you start with a base layer and then add in what you need.
+In many ways, building a container is like installing software on your computer, except that you don't have an interactive prompt.
+If you can install software on your computer from the command line then you can build a docker container.
+
+A few things to note:
+- Containers are built in layers, with each layer being a container of it's own.
+- It is often best to start with a container that does most of what you want rather than something "empty" like `Ubuntu`.
+- Whatever files you put into the container during the build process will remain there during deployment.
+
+Docker containers are built according to instructions in a `Dockerfile`.
+An example is shown below:
```

Comment:
Probably want to say it's an example for Robbie and include a github link

Reply:
Good idea - done.